### DISCUSSION TITLE
Blog posts in folders (with assets)/nested shortcodes

### DISCUSSION THREAD

**Discussion by Brixy** [lumeland/lume#197] [Q&A] [answered]
Hi all!

After playing with Lume I am blown away by its flexibility. Excellent work, guys!

I would like to migrate a number of rather large scale Hugo and Eleventy projects to Lume and I have a few questions:

- Is it/how is it possible to use a Hugo-like folder structure for a blog?
- Is it possible to copy assets (and process images with Imagick) from within these folders?

```
## In

blog/
├─ 2022-05-01_first-post/
│  ├─ index.md
│  ├─ an-image.png
│  └─ a-pdf-file.pdf
├─ 2022-05-02_second-post.md
└─ 2022-05-03_third-post/
   ├─ index.md
   └─ another-image.png

## Out

blog/
├─ first-post/
│  ├─ index.html
│  └─ an-image.png
├─ second-post.html
└─ third-post/
   ├─ index.html
   └─ another-image.png

```

- Is there an equivalent of Eleventy’s *paired shortcodes.* This is an easy example of what I would like to do:

```
site.addHelper('blockquote', blockquote, { body: 'true' });

{% blockquote author="XY", link="https…" %}
This is a quote formatted with **Markdown**.

Another line of text.
{% endblockquote %}

Thank you very much!

**Comment by oscarotero** [comment, marked as answer]
Hi, @Brixy 

> Is it/how is it possible to use a Hugo-like folder structure for a blog?

The url generation in Lume doesn't work exactly like Hugo: The date from the path is extracted only if it's in the filename but not in the folder name. In your example, `/blog/2022-05-01_first-post/index.md` outputs to `/blog/2022-05-01_first-post/index.html`, but `blog/2022-05-02_second-post.md` outputs to `blog/second-post/` (because the date is defined in the filename, not in the folder). I think this could be implemented easily but it's not available right now.

Fortunately, you can use javascript functions to customize the final url of the pages [See this](https://lume.land/docs/creating-pages/urls/#urls-as-functions). You can create a `blog/_data.js` file and export the `url` function that will be used to generate the url of the post as you like:

```js
export function url(page) {
    const path = page.src.path; // The path name without extension (ex: /blog/2022-05-01_first-post/index)
    const url = generateMyOwnUrl(path); // This returns something like /blog/first-post/
    return url;
}
```

Another way is by creating a preprocessor, that can change the url before rendering the page:

```js
// Preprocess all html pages
site.preprocess([".html"], (page) => {
    const path = page.src.path; // The path name without extension (ex: /blog/2022-05-01_first-post/index)
    const url = generateMyOwnUrl(path); // This returns something like /blog/first-post/
    page.url = url; // Assign the new url to the page
});
```

> Is it possible to copy assets (and process images with Imagick) from within these folders?

Yes. If you are going to use imagick plugin, just import it in your config file and that's all. By default the extensions `.jpg`, `.jpeg` and `.png` are loaded but you can change it.

```js
import lume from "lume/mod.ts";
import imagick from "lume/plugins/imagick.ts";

export default lume()
    .use(imagick({
        extensions: [".jpg"], //If you only want to load jpg files
    ));
```


> Is there an equivalent of Eleventy’s paired shortcodes.

Yes, you have to define the helper as `type: "tag"` ([See this](https://lume.land/docs/configuration/filters/#helpers)). In your example:

```js
site.addHelper('blockquote', blockquote, { body: true, type: "tag" });
```

**Reply by Brixy** [reply]
Gosh! It seems that all my wishes I have had for Eleventy and Hugo for years have come true with Lume.

Thank you very much @oscarotero for your time and your excellent explanation!

Currently, my solution is rather primitive, but it works just fine:

```
export function url(page) {
  url =
    page.src.path
      .replace('/posts', '')
      .replace(/[0-9]{4}-[0-9]{2}-[0-9]{2}_/, '') +
    page.src.ext.replace('.md', '.html');
  return url;
}
```

Being new to Deno I have one last questions (only if ok!):

As I have to use a `_data.js` file is it possible to add Imagick settings in the `.js` file, e.g. like this:

```
export const imagick = {
  format: 'webp'
}
```

Thank you very much. Really appreciate your work.

**Reply by oscarotero** [reply]
Great!
In your url function, I think you don't need to replace the extension (is always `.md`) so I'd change to something like this:

```js
export function url(page) {
  return page.src.path
      .replace('/posts', '')
      .replace(/[0-9]{4}-[0-9]{2}-[0-9]{2}_/, '') + ".html";
}
```

And about the `_data.js`, yes you can export the imagick settings here.

**Reply by Brixy** [reply]
Thanks!

I changed the extension because the `.jpg` files were changed into `.html` files, too.

Moreover, I noticed that images seem to be treated like markdown files if they live in a folder structure like posted above. Their source is replaced by my `base.njk`’s HTML code – even if I don’t manipulate the pages’ paths at all.

So, I need to play a bit more ;-)

**Reply by oscarotero** [reply]
yes, in Lume all files are "pages", not only text files like markdown or nunjucks, but also images, etc, so any info that you place inside a _data.* file (like the url or imagick settings) is shared to all pages inside this directory. This affects also to configurations like `layout`.

As a workaround, you can remove the layout value of the images with this preprocessor:

```js
site.preprocess([".jpg"], (page) => page.data.layout = undefined);
```

**Reply by Brixy** [reply]
Excellent! Thank you very much again.

Everything seems to be working now.

It seems I still have to use

```
…
page.src.ext.replace('.md', '.html');
…
```

```
.replace(/[0-9]{4}-[0-9]{2}-[0-9]{2}_/, '') + '.html'
```

also renames image files.

Thank for your excellent SSG and your top support!

**Comment by oscarotero** [comment]
I've created an issue to support dates in directories: https://github.com/lumeland/lume/issues/198

But I have some questions?:
- It's required the file name to be "index.md" (`posts/2022-05-01_first-post/index.md`) or it works with any filename? (like `posts/2022-05-01_first-post/first-post.md`)
- Does it works if the date is in a parent directory? (like `posts/2022-05-01_first-post/first-post/index.md`)

This week I'd like to release a new Lume version and want to include this.
Thanks!

**Reply by Brixy** [reply]
Great! Really looking forward.

My subjective answers:

> It's required the file name to be "index.md"

Hugo uses `posts/2022-05-01_first-post/index.md`, and eleventy `posts/first-post/2022-05-01-index.md` (but without image processing from content folders, so it does not make too much sense here).

So, I would suggest `index.md`, also because renaming can be tricky if there is a duplication in the file and folder name.

> Does it works if the date is in a parent directory?

I think nested folders with redundant names are rather inconvenient, too.

With `posts/2022-05-01_first-post/index.md` people coming from Hugo, Jekyll and others SSGs only would have to replace a single `-` with `_` – and everything just works.

Thank you very, very much!

**Reply by Brixy** [reply]
One more thought concerning your new issue:

It is extremely convenient to store all page-related assets in one folder.

If I understood it correctly this currently only works for images (which is fantastic and the most important use case by far).

Would it make sense to support a structure like this?

```
posts/
└─ 2022-05-01_first-post/
   ├─ index.md
   ├─ file.pdf
   ├─ audio.mp3
   ├─ video.mp4
   └─ an-image.png
```

**Reply by oscarotero** [reply]
Great, thanks!

To load pdf, mp3 or any other file type, just configure Lume to load them:

```js
import lume from "lume/mod.ts";
import binaryLoader from "lume/core/loaders/binary.ts";

const site = lume();

site.loadAssets([".pdf", ".mp3", ".mp4"], binaryLoader);

export default site;
```

This tells Lume to load files with these extensions (in the same way that loads images). The binaryLoader is required here, because by default it loads the files as text files. Note that this will read the content of the files, so it affect to the memory used.

In fact, the imagick plugin does that: it configure Lume to load some files (jpg, and png) and then register a processor to modify them.

There's a way to copy files without read them, if you don't need to process (which is faster), with `site.copy()` but it cannot be configured by extension. You must set specific files or folders to copy. This is something l'm planning to improve but it's not there yet.

**Reply by Brixy** [reply]
Wow. Unbelievable how easy all this is with Lume.

Thanks for your snippet!

I will transfer all websites to Lume.

**Reply by oscarotero** [reply]
The next version of Lume (1.9.0) will include the ability to extract dates from directories, and copy files by extension.
In this example:

```
posts/
└─ 2022-05-01_first-post/
   ├─ index.md
   ├─ file.pdf
   ├─ audio.mp3
   ├─ video.mp4
   └─ an-image.png
```

And this configuration:

```ts
import lume from "lume/mod.ts";

export default lume()
  .copy([".pdf", ".mp3", ".mp4"]);
```

The output is:

```
posts/
└─ first-post/
   ├─ index.html
   ├─ file.pdf
   ├─ audio.mp3
   ├─ video.mp4
   └─ an-image.png
```

**Reply by Brixy** [reply]
Fantastic. Looking forward.

For our use cases Lume is just the best SSG by far. The switch was a breeze. I could just re-use all our Eleventy shortcodes and Nunjucks templates.

Thank you very much for your efforts, and for Lume, and your excellent support!

**Reply by oscarotero** [reply]
Nice. FYI, you can run `lume upgrade --dev` to install the last development version (the last commit) and test this new features.

**Reply by Brixy** [reply]
The development version works perfectly. Thanks a lot!

Lume even accepts static `.html` files:

```
site
.copy('static', '.')
.copy(['.html', '.mp3', '.mp4', '.pdf']);
```

I just noticed one thing:

When you manipulate the output path (via `/src/posts/_data.js`) like this

```
export function url(page) {
  return (
    page.src.path
      .replace('/posts', '')
  );
}
```

- posts from folders (e.g. `/src/posts/2022-05-14_post-1/`) are exported to the expected folder, e.g. `/_site/post-1/`,
- posts from files (e.g. `/src/posts/2022-05-14_post-2.md`) are copied to `/_site/2022-05-14_post-2` (without extension),
- static files are copied to e.g. `/site/posts/post-1/video.mp4`.

Is this intended? I will research further, but wanted to inform you already.

**Edit:** Another note: `imagick` works with this setup. I will try to add the `copy` function to `/src/posts/_data.js`.

**Reply by oscarotero** [reply]
Yes, it's intended. The `url` function only works for loaded pages that need to be processed (like markdonw files or jpg files that are processed with imagick). `copy()` function doesn't  load the files, because they are not processed, only copy them as is.
Maybe I could implement something like this:

```ts
site.copy(['.html', '.mp3', '.mp4', '.pdf'], (path) => path.replace("/posts", ""));
```

This allows to use a function as the second argument that will be used to resolve the final url of every file copied.

**Reply by Brixy** [reply]
OK, I understand. Thanks for pointing this out.

> Maybe I could implement something like this:
>
> `site.copy(['.html', '.mp3', '.mp4', '.pdf'], (path) => path.replace("/posts", ""));`

That would be totally great (if this is of general interest). So one can just use `[download PDF](file.pdf)` in one’s markdown code.

Repeating `.repeat` as a second argument would make multiple collections/blogs with static files possible so you can use `/posts/`, `/notes/`, `/whatever/` etc.

(By the way: I would really like to help here, but I am not a programmer but just a passionate user.)

**Reply by oscarotero** [reply]
Ok, no probem, this is something that I had planned to implement (and I just did). Please, can you update your lume dev version (run `lume upgrade --dev` again) and test it.

**Reply by Brixy** [reply]
Thanks once more!

I don’t exactly know how everything works together.

Here is what I get:

## Without any `url(page)` replacement and without secondary copy argument

Everything just works! This is probably the most common use case.

## Socondary argument for site.copy

```
site
.copy('static', '.')
.copy(['.docx', '.mp3', '.mp4', '.pdf'], (path) => path.replace("/posts", ""));
```

Assets are copied to the correct place (`/_site/post-1/`).

Posts and images are in `/_site/posts/…/` as expected.

## With url replacement

With this in `/src/posts/_data.js`:

```
export function url(page) {
  return (
    page.src.path
    .replace('/posts', '')
    .replace(/[0-9]{4}-[0-9]{2}-[0-9]{2}_/, '') + '.html' // this seems to be necessary even with the new get date from folders functionality
  );
}
```

the behaviour seems a bit strange to me:

- `index.html` and assets are located in the correct folders, e.g. `/_site/post-1/`,
- some images (generated with `imagick`) are located in `/_site/post-1/`, others in `/_site/posts/post-1/`.

(I deleted the `/_site/` each time.)

**Reply by oscarotero** [reply]
The purpose of these changes is that you don't need to set a `url` function for that, so you can remove it.
Regarding to the `url` function itself, the `page.src` property contain info about the source file of the page (file name, extension, etc). There's the `page.dest` property with info about the output destination, so this function could be simply:

```js
export function url(page) {
  return page.dest.path.replace("/posts", "") + ".html";
}
```

**Reply by Brixy** [reply]
Yes, `page.dest.path` is the solution.

Date and assets handling is incredibly easy now – hopefully for all people switching to Lume from Hugo et al.

THANK YOU VERY, VERY MUCH for your help and patience!

**Reply by oscarotero** [reply]
Thank you for give a try to Lume. If you create something interesting with it, please let me know so I can add it to the showcase section: https://lume.land/showcase/

**Reply by Brixy** [reply]
Yes, will certainly do ;-)

**Comment by Brixy** [comment]
The combination of removing the `/posts` folder from path + using generating posts from folders and markdown files (`/src/posts/post-1/` and `/src/post-2.md`) + using `imagick` with all images and assets living in the posts’ folders seems a bit tricky.

So for reference, this is what I have come up with – in case anyone might need it.

There are probably more elegant solutions; but this works ;-)

```
# _config.ts

site
  .copy('static', '.')
  .copy(
    ['.docx', '.mp3', '.mp4', '.pdf'],
    (path) => path.replace('/posts', ''),
  );

site.preprocess(['.jpg', '.png', '.gif'], function (page) {
  page.data.layout = undefined;
});

site.process(['.jpg', '.png', '.gif'], function (page) {
  page.dest.path = page.dest.path.replace('/posts', '');
});


# /src/posts/_data.yml

export function url(page) {
  let path = page.dest.path.replace('/posts', '');
  if (!path.includes('/index') && page.src.ext == '.md') {
    path = path + '/index';
  }
  return path + page.src.ext.replace('.md', '.html');
}

export const imagick = [
  {
    resize: [200, 200],
    suffix: '-small',
  },
  {
    resize: [200, 200],
    suffix: '-small',
    format: 'webp',
  },
  {
    resize: [600, 600],
  },
  {
    resize: [600, 600],
    format: 'webp',
  },
];
```

**Reply by oscarotero** [reply]
This is a more simple solution (I didn't test it, but can help as an example):

```js
// _config.ts

site
  .copy('static', '.')
  .copy(
    ['.docx', '.mp3', '.mp4', '.pdf'],
    (path) => path.replace('/posts', ''),
  );


// /src/posts/_data.yml

export function url(page) {
  const path = page.dest.path.replace("/posts", "");

  return page.dest.ext === '.md'
    ? path + "/"
    : path + page.dest.ext;
}

export const imagick = [
  {
    resize: [200, 200],
    suffix: '-small',
  },
  {
    resize: [200, 200],
    suffix: '-small',
    format: 'webp',
  },
  {
    resize: [600, 600],
  },
  {
    resize: [600, 600],
    format: 'webp',
  },
];
```

- The `url` function is applied to any file loaded in this directory (not only markdown but also the images processed with imagick), so there's no need to create a processor for that.
- The preprocessor to remove the layout value of the images is a recurrent issue, so I decided to change Lume to ignore it automatically. This will be available in the next version but if you are using the development version, you can remove this preprocessor too.
- If the url returned by `url` function ends with a slash (like `/first-post/`) Lume automatically append the `index.html` to the end.

Let me know if this works for you.

**Reply by Brixy** [reply]
Thanks for pointing me to the ternary operator. Works perfectly.

I just had to get rid of `/index/` subfolders:

```
export function url(page) {
  const path = page.dest.path.replace("/posts", "").replace("/index", "");

  return page.dest.ext === ".md"
    ? path + "/index.html"
    : path + page.dest.ext;
}
```

Thanks a lot for your excellent 1.9 version!

**Reply by Brixy** [reply]
Last tiny addition ;-)

Without

```
site.process(['.jpg', '.png', '.gif'], function (page) {
  page.dest.path = page.dest.path.replace('/posts', '');
});
```

*some* of the images generated by `imagick` are placed in `/posts/post-1/` etc.

**Reply by oscarotero** [reply]
> some of the images generated by imagick are placed in /posts/post-1/ etc.

Mmm, this may be a bug. I'll take a look

**Reply by oscarotero** [reply]
I can't reproduce this issue. Do you know when this happens?

**Reply by Brixy** [reply]
Thanks for checking this!

Although it’s probably not a Lume issue I just pushed a [minimal example](https://github.com/Brixy/lume-minimal-example), and this is the folder structure Lume produces:

```
🔥 / /index.md
🔥 /post-3/ /posts/2022-01-03_post-3.md
🔥 /post-2/ /posts/2022-01-02_post-2/index.md
🔥 /post-1/ /posts/2022-01-01_post-1/index.md
🔥 /posts/post-1/test-small.jpg /posts/2022-01-01_post-1/test[0].jpg
🔥 /posts/post-1/test-small.webp /posts/2022-01-01_post-1/test[1].jpg
🔥 /post-1/test.webp /posts/2022-01-01_post-1/test.jpg
🔥 /posts/post-1/test.jpg /posts/2022-01-01_post-1/test[2].jpg
```

If the `site.process` code in `_config.ts` is activated the folder structure looks like this:

```
🔥 / /index.md
🔥 /post-3/ /posts/2022-01-03_post-3.md
🔥 /post-2/ /posts/2022-01-02_post-2/index.md
🔥 /post-1/test.webp /posts/2022-01-01_post-1/test.jpg
🔥 /post-1/ /posts/2022-01-01_post-1/index.md
🔥 /post-1/test-small.jpg /posts/2022-01-01_post-1/test[0].jpg
🔥 /post-1/test-small.webp /posts/2022-01-01_post-1/test[1].jpg
🔥 /post-1/test.jpg /posts/2022-01-01_post-1/test[2].jpg
```

**Reply by oscarotero** [reply]
Nice. Found the bug. It will be fixed in v1.9.1
Thanks!
